### ISSUE TITLE
Error 500 gives blank page with Lumen

### ISSUE THREAD

**Issue by jonaholsson** [dingo/api#414]
Using the latest dev build to support Lumen, I'm getting blank error 500 pages. Am I missing any configuration parameter? Debug is set to true in both Dingo config and Lumen config.

**Comment by catalinux** [comment]
Please check the log files. It might not take your config variable for debug. 500 error means that a php error happened.

make sure you have 
```php
Dotenv::load(__DIR__ . '/../'); 
```
in app.php and also 

APP_ENV=local
APP_DEBUG=true

in your .env file (not .env-sample)

**Comment by jonaholsson** [comment]
Thanks @catalinux, I found the issues in the log files, but both app.php and .env are correct.

**Comment by catalinux** [comment]
What errors do you have?

**Comment by jonaholsson** [comment]
They are all various PHP errors, such as missing controllers and other PHP syntax errors, but whenever I'm trying a missing route the Dingo 404 error response is shown correctly.

**Comment by jasonlewis** [comment]
Do you have a prefix and the routes setup correctly to handle the prefix? Paste your routes and relevant `API_*` definitions from your `.env` file.

**Comment by jonaholsson** [comment]
Here are my `API_*` definitions from `.env`:

```API_VENDOR=testapp
API_VERSION=v1
API_PREFIX=api
API_CONDITIONAL_REQUEST=true
API_STRICT=false
API_DEBUG=true
API_DEFAULT_FORMAT=json```

My one and only route at the moment:

```$api->version('v1', ['prefix' => 'api'], function ($api) {
	$api->post('test', 'App\Http\Controllers\TestController@postTest');
});```

**Comment by slovenianGooner** [comment]
I have the same issue, yes.

**Comment by jasonlewis** [comment]
Fresh Lumen install? Anything in the logs. I'm not sure what could be causing issues. The only `API_*` definitions you really need are the vendor and prefix (or domain if you're on a different domain).

Please provide any relevant error logs and code.

**Comment by slovenianGooner** [comment]
I've basically tracked down the issue to here:

```
    public function render($request, Exception $exception)
    {
        if ($request instanceof Request) {
            throw $exception;
        }

        return $this->parent->render($request, $exception);
    }
```

in the ```Dingo\Api\Exception\Handler``` class.

The request is an instanceof Request and therefore throws another Exception, presuming it's going to be handled by the "normal" exception handler? At least that's my understanding. And then I get lost on where the exception is handled. It gets written in the lumen.log file, but doesn't display anything bar a blank page.

**Comment by jasonlewis** [comment]
Can you track where the exception is being thrown from? Is it during the dispatching of the route or prior to this?

**Comment by slovenianGooner** [comment]
It's after dispatching the route, basically any error in the controllers, etc. throws a blank page. It's a PDOException in this speific case, but also throwing a general Exception manually returns a blank page.

**Comment by jasonlewis** [comment]
Okay that helps, I'll take a look but it should be be handled by the handler. It's a little annoying because Lumen handles exceptions differently to Laravel.

**Comment by slovenianGooner** [comment]
Yeah, I get that, that's why I thought it was an issue with Lumen. Just tested and disabling the Dingo package and registering a generic route returns a "good" error. :)

**Comment by jasonlewis** [comment]
I've tweaked the exception handler so it now handles all exceptions by default and will build a generic response if there's no custom handler registered. It will default to a `500 Internal Server Error` for the response type as well. The exceptions message (if any is provided) will be used as the message key.

**Comment by slovenianGooner** [comment]
Thanks for this, mate! Much appreciated! Will test it and get back if there are any issues.

**Comment by jasonlewis** [comment]
Awesome much appreciated.

**Comment by slovenianGooner** [comment]
Hm, still getting a completely blank page. Will debug a bit and see where it's stuck.

**Comment by jasonlewis** [comment]
What kind of exception and where are you throwing it?

**Comment by slovenianGooner** [comment]
This, for example returns a JSON error with the Dingo API Exception handler:

```
$api = app()['api.router'];

$api->version('v1', ['protected' => false], function($api) {

    $api->get('/', function(){
        throw new Exception("Test");
        return 'a';
    });

});
```

Then a fatal PHP error returns a blank page, for example from this route:

```
$api->get('/', function(){
        $user = User::first();
        return 'a';
    });
```

This just examples, but the class User cannot be found up there and it gives a fatal PHP error, resulting in a blank page. If I turn on PHP errors, it gives a full stack of course.

**Comment by jasonlewis** [comment]
Right of course. I'll take another look.

**Comment by jasonlewis** [comment]
I tried your example on a fresh Lumen install with `display_errors=Off` in `php.ini` and I get a blank screen.

**Comment by slovenianGooner** [comment]
So must be the Lumen then? When I disable the Dingo API service provider I get a perfectly fine error page, that's what baffles me.

**Comment by jasonlewis** [comment]
Can you screenshot the error page for me?

**Comment by slovenianGooner** [comment]
The default one without the provider enabled?

**Comment by jasonlewis** [comment]
Yeah and display_errors off I'm guessing?

**Comment by slovenianGooner** [comment]
![routenotfound](https://cloud.githubusercontent.com/assets/1257629/7859913/f37818dc-0544-11e5-8106-3851dfa70ca8.PNG)

There you got. Basic 404 exception without the service provider and display errors set to off.

**Comment by jasonlewis** [comment]
Things like syntax errors and things working correctly? Using a class that
doesn't exist?

**Comment by slovenianGooner** [comment]
![syntaxerror](https://cloud.githubusercontent.com/assets/1257629/7859995/9bd289cc-0545-11e5-99ac-ee5deef961da.PNG)

Yes, seems to.

```
$app->get('/', function(){
   return asdas;
});
```

**Comment by slovenianGooner** [comment]
Just did another composer update. Syntax violation seems to work with a "Dingo" now, but undefined class still returns a blank page.

**Comment by jasonlewis** [comment]
Okay no worries I'm going to have to play with this a bit more it seems.
Bloody Lumen doing things differently! Heh.

**Comment by slovenianGooner** [comment]
Yeah, no problem. Wish I could help out, if I find something while having display_errors on I'll let you know.

**Comment by jasonlewis** [comment]
I'm going to re-open this. I'll have to investigate some more into what the hell Lumen is doing.

**Comment by jonaholsson** [comment]
Thanks @jasonlewis. I'll dig a little bit further too.

**Comment by masnug** [comment]
500 gives blank page is happen in laravel too @jasonlewis

**Comment by jasonlewis** [comment]
Sample code producing a blank screen?

**Comment by masnug** [comment]
```php
    public function me()
    {
        $user = $this->user();

        $response = $this->response->item($user, new UserTransformer);

        return $response;
    }
```
UserTransformer is throwing exception, but its just gave 500 blank page

**Comment by jasonlewis** [comment]
I've fixed a couple issues surrounding some WSOD with Laravel.

As for Lumen, the issue is that there's no shutdown handler for fatal errors. I've opened up laravel/lumen-framework/pull/188 for Taylor to have a look at. This should resolve those annoying WSOD when `display_errors` is off.

**Comment by jasonlewis** [comment]
The PR has been merged into the 5.1 branch so as yet there's no stable tag. Use `5.1.*@dev` to get the latest.

I'll keep this open until there's a stable tag on Lumen.

**Comment by jasonlewis** [comment]
Lumen `5.1.3` has been tagged and contains the latest changes to the exception handling so all should be sweet now. Hopefully no more WSOD!
